tests/raftstore/transport_simulate.rs

```diff
@@ -45,51 +47,58 @@ struct FilterDelay {
 struct FilterOutOfOrder;
 
 impl Filter for FilterDropPacket {
-    fn before(&self, _: &RaftMessage) -> bool {
-        rand::random::<u32>() % 100u32 < self.rate
+    fn before(&mut self, _: &RaftMessage) -> bool {
+        self.drop = rand::random::<u32>() % 100u32 < self.rate;
+        self.drop
     }
-    fn after(&self, x: Result<()>) -> Result<()> {
+    fn after(&mut self, x: Result<()>) -> Result<()> {
+        if self.drop {
+            return Err(Error::Timeout("make by FilterDropPacket in SimulateTransport".to_string()));
+        }
         x
     }
 }
 
 impl Filter for FilterDelay {
-    fn before(&self, _: &RaftMessage) -> bool {
+    fn before(&mut self, _: &RaftMessage) -> bool {
         sleep_ms(self.duration);
         false
     }
-    fn after(&self, x: Result<()>) -> Result<()> {
+    fn after(&mut self, x: Result<()>) -> Result<()> {
         x
     }
 }
 
 impl Filter for FilterOutOfOrder {
-    fn before(&self, _: &RaftMessage) -> bool {
+    fn before(&mut self, _: &RaftMessage) -> bool {
         unimplemented!()
     }
-    fn after(&self, _: Result<()>) -> Result<()> {
+    fn after(&mut self, _: Result<()>) -> Result<()> {
         unimplemented!()
     }
 }
 
 pub struct SimulateTransport<T: Transport> {
-    filters: Vec<Box<Filter>>,
+    filters: Vec<RwLock<Box<Filter>>>,
```

Comment:
Because @siddontang don't want trait Transport has a send(&mut self, ...) mutable interface, while SimulateTransport want to change the inner state when it call send. 
Maybe we can get better solution for that , but I'm not sure.